Library/Homebrew/sbom.rb

```diff
@@ -0,0 +1,354 @@
+                :changed_files, :poured_from_bottle, :loaded_from_api, :time, :stdlib, :aliases, :arch, :source,
+                :built_on, :license, :name
+  attr_writer :compiler, :runtime_dependencies, :source_modified_time
+
+  # Instantiates a {SBOM} for a new installation of a formula.
+  sig { params(formula: Formula, compiler: T.nilable(String), stdlib: T.nilable(String)).returns(T.attached_class) }
+  def self.create(formula, compiler = nil, stdlib = nil)
+    build = formula.build
+    runtime_deps = formula.runtime_dependencies(undeclared: false)
+    attributes = {
+      "name"                    => formula.name,
+      "homebrew_version"        => HOMEBREW_VERSION,
+      "spdxfile"                => formula.prefix/FILENAME,
+      "built_as_bottle"         => build.bottle?,
+      "installed_as_dependency" => false,
+      "installed_on_request"    => false,
+      "poured_from_bottle"      => false,
+      "loaded_from_api"         => false,
+      "time"                    => Time.now.to_i,
+      "source_modified_time"    => formula.source_modified_time.to_i,
+      "compiler"                => compiler,
+      "stdlib"                  => stdlib,
+      "aliases"                 => formula.aliases,
+      "runtime_dependencies"    => SBOM.runtime_deps_hash(formula, runtime_deps),
+      "arch"                    => Hardware::CPU.arch,
+      "license"                 => SPDX.license_expression_to_string(formula.license),
+      "built_on"                => DevelopmentTools.build_system_info,
+      "source"                  => {
+        "path"         => formula.specified_path.to_s,
+        "tap"          => formula.tap&.name,
+        "tap_git_head" => nil, # Filled in later if possible
+        "spec"         => formula.active_spec_sym.to_s,
+        "patches"      => formula.stable&.patches,
+        "bottle"       => formula.bottle_hash,
+        "stable"       => {
+          "version"  => formula.stable&.version,
+          "url"      => formula.stable&.url,
+          "checksum" => formula.stable&.checksum,
+        },
+      },
+    }
+
+    # We can only get `tap_git_head` if the tap is installed locally
+    attributes["source"]["tap_git_head"] = T.must(formula.tap).git_head if formula.tap&.installed?
+
+    new(attributes)
+  end
+
+  sig { params(attributes: Hash).void }
+  def initialize(attributes = {})
+    attributes.each { |key, value| instance_variable_set(:"@#{key}", value) }
+  end
+
+  sig { returns(T::Boolean) }
+  def valid?
+    data = to_spdx_sbom
+
+    schema_string, _, status = Utils::Curl.curl_output(SCHEMA)
+
+    opoo "Failed to fetch schema!" unless status.success?
+
+    Homebrew.install_bundler_gems!(groups: ["bottle"])
+    require "json_schemer"
+
+    schemer = JSONSchemer.schema(schema_string)
+
+    return true if schemer.valid?(data)
+
+    opoo "SBOM is not valid, not writing to disk!"
```

Comment:
Should this maybe `odie` (or `raise` after printing errors) so we figure out these problems and fix them?

Reply:
Maybe die if an environment variable is set or something? I guess I'm just thinking for now "how will we know if this ever fails?"

Reply:
Yeah, good point. `HOMEBREW_ENFORCE_SBOM`?

Reply:
@SMillerDev Yeh. Could set it unconditionally in homebrew-core CI and make it optional/dependent on label if it breaks on individual formulae we need to override.

Reply:
I think if it breaks on a formula that's just something we need to fix here. I can't really think of a reason generating this metadata would fail for one and not another formula

Reply:
@SMillerDev it feels like the output is gonna be very formulae dependent? see the example below you mentioned about e.g. patches being present/not.

Reply:
@SMillerDev the "should" here does not make me feel overly confident but happy to just merge and can revert if needed.

Reply:
So, I'm very certain that no amount of entries in the arrays will break the schema. I'm not so certain that my code will work correctly with all variations. So I'm expecting any issues to be fixable here, and not require formula specific overrides.